### ISSUE TITLE
Support RCS API (if available)

### ISSUE THREAD

**Issue by miladmaz12** [moezbhatti/qksms#1522]
### DESCRIPTION
A short description of the issue. 
Example: Conversations aren't marked as read when they are opened from a notification.

### STEPS
1. Steps
2. to 
3. reproduce
4. the 
5. issue.

### EXPECTED
Describe the expected behaviour.

### OBSERVATIONS
Describe what happens instead of the expected behaviour.

**Comment by mkeedlinger** [comment]
I'll add a little bit of context since the body of this issue doesn't have much.

[RCS](https://en.wikipedia.org/wiki/Rich_Communication_Services) brings iMessage like functionality to Android in a cross-carrier, standardized fashion. The protocol and efforts to bring support for it have been spear-headed by Google, and [in a recent post](https://www.blog.google/products/rcs/upgrading-messaging-android-us-rcs/) they announced RCS would be rolled out soon, but that it is currently mainly supported by their Messages app.

RCS is awesome, having support for it in an OSS implementation would be incredible.

<sup>Aside: Thanks for this app, it's the only real OSS messaging alternative I can find on fdroid</sup>

**Comment by e-t-l** [comment]
One thing to consider is the switch between SMS/RCS. AFAIK, Google Fi is the only carrier that doesn't charge data usage for RCS messages, so it might be handy to have an option that enables RCS only over wifi, for users who don't want to eat up their data allowance. It's also the responsibility of the SMS app to automatically re-send RCS messages as SMS if the phone doesn't have a data connection. Android Messages already has the latter feature (although I haven't been able to test it out), but I think implementation of the former would be an excellent feature to include in an OSS app.

**Comment by moezbhatti** [comment]
I'd love to implement it as soon as it's possible to, but unfortunately we're still waiting on the public API from Google

**Comment by AyoungDukie** [comment]
Just to bump, it looks like RCS has *started to be implemented outside of Google Messages, at least a couple apps in the Play Store and apparently the OnePlus messaging app

**Comment by iot-resister** [comment]
https://github.com/android-rcs/rcsjta

> Thanks to its client/server API, the stack may be easily integrated with existing native Android applications (e.g. address book, dialer) and permits to create new RCS applications (e.g. chat, widgets).

**Comment by eternaltyro** [comment]
This will be an awesome feature to have. I'll be willing to chip in a little bit to sponsor this feature as well; as I'm sure others using the app also will.

**Comment by phpbg** [comment]
> I'd love to implement it as soon as it's possible to, but unfortunately we're still waiting on the public API from Google

RCS is supposed to be a standard.
Can't it be implemented from scratch?

**Comment by iot-resister** [comment]
@phpbg 
https://github.com/android-rcs/rcsjta

**Comment by eternaltyro** [comment]
@phpbg  We should not have to implement it from scratch because maintaining it would be difficult. Existing implementations used by multiple products will be subject to more scrutiny and therefore (intuitively) might be better in terms of performance and security. So It's better to use a good third-party library with permissive license.

**Comment by phpbg** [comment]
> @phpbg We should not have to implement it from scratch because maintaining it would be difficult. Existing implementations used by multiple products will be subject to more scrutiny and therefore (intuitively) might be better in terms of performance and security. So It's better to use a good third-party library with permissive license.

Agree, but I cannot find any actively maintained and open source library... :-/

**Comment by ValZapod** [comment]
> RCS is supposed to be a standard.

No. The standard is too hard. https://www.gsma.com/futurenetworks/resources/rcs-up-2-4-uni/

And that is just universal profile, I am not talking about E2EE of Google.

**Comment by freshgum-bubbles** [comment]
@iot-resister that library you sent seems to be abandoned... are there any viable boil-in-a-bag client library alternatives?

It looks a bit like an open-but-closed Google standard. The fact that we haven't had a decent maintained library for it in over 4 years should raise eyebrows, and perhaps hint at the complexity behind RCS, or some sort of intention to keep it out of an open-source library, as is the usual way with Google.

I doubt RCS is coming to QKSMS soon...

**Comment by ValZapod** [comment]
> open-but-closed Google standard.

It is mandatery part of 5g/LTE. Just like ipv6. So no. Indeed google has its own extensions.

> at the complexity behind RCS,

It MUST be part of OS. Google is still patching it right now. https://android-review.googlesource.com/q/RCS But then where is 255 segment SMS support in Android itself? Right, no support, it converts to MMS!

**Comment by jtagcat** [comment]
jfyiya: RCS and RCS are different things. One is Google's, other is generic.

### GitHub Etiquette

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on new comments, or get notified of resolution with 'Status'.

**Comment by jtw023** [comment]
> Can I sponsor this feature?

I would sponsor this feature in a heartbeat! Be sure to push a notification out alerting people when an OSS implementation of this becomes a real possiblity. I'm sure a large portion of your user base would be willing to sponsor as well.

**Comment by markcellus** [comment]
I'd sponsor this also, for sure 👍

**Comment by Tobias-Lauerbach** [comment]
> I'd love to implement it as soon as it's possible to, but unfortunately we're still waiting on the public API from Google

I hve found this APP which looks nearly identical (possibpy a fork) and sas it woud support rcs but i coudnt confirm it..., but mayby it helps
https://github.com/vidit135g/Replify-Messenger

@moezbhatti

**Comment by Katherine1** [comment]
> > I'd love to implement it as soon as it's possible to, but unfortunately we're still waiting on the public API from Google
> 
> I hve found this APP which looks nearly identical (possibpy a fork) and sas it woud support rcs but i coudnt confirm it..., but mayby it helps https://github.com/vidit135g/Replify-Messenger
> 
> @moezbhatti

Seeing as this is still an open issue, yeah, at first glance, that's looking like a fork. It's even giving QKSMS credit at the bottom. If that project has RCS implemented as it claims, backporting that work should be doable? I dunno. I think I'll grab the QKSMS code and try getting familiar with it tonight and go from there.

**Comment by markcellus** [comment]
I took a look at https://github.com/vidit135g/Replify-Messenger. I wouldn't recommend people use it.

1. The code in the repository doesn't appear to be integrating RCS, so saying it does in the README isn't truthful
2. The only edit that I can see after the initial code was copied into this repo is a change to the README and a credit to QKSMS. This is also a little disingenuous because not forking the QKSMS repository makes it difficult to see any diffs or specific edits this maintainer has made to the QKSMS code before copying it.
3. The app in the Google Play store shows that it's owned by a company called "The Absolute Corporation" with a gmail address and has what appears to be a fake business address.

These are just a few of the red flags and time is limited to list them all, but they are enough for me to stay far away from it. Additionally, the feature set seems to be a full copy of the QKSMS and the last release is years older than QKSMS. There's no reason to use this Replify-Messenger app.

**Comment by ilmaisin** [comment]
https://github.com/Hirohumi/rust-rcs-core

Someone seems to be working on a RCS client library.

The bigger question is: which server it could be used to connect to? Very few carriers provide RCS servers of their own, and many have actually shut down their ones in favour of Google Jibe.

I am not sure what RCS in its current iteration means to AOSP phones, or things like Pinephone and Librem 5? Will corporations such as Google, Apple and Huawei be the only ones that can build modern texting apps?

**Comment by YannBirba** [comment]
Do you have any news about RCS support in qk ?

**Comment by onyxiarivera** [comment]
Would love to see this added to QK, I absolutely despise Google Messages and now they're blocking RCS for rooted/custom rom installs of Android.

**Comment by trymeouteh** [comment]
Would like to see fully feature RCS support

**Comment by lascapi** [comment]
Hi there! 
I'm not sure if the Android API for using RCS is open yet. 
So I don't know if it's possible to implement RCS in quik. 
It's a difficult question to work on. 

Here is my question: how much money do we want to pay to hire an engineer (software + telecom) to work on it?

**Comment by ilmaisin** [comment]
> Hi there! I'm not sure if the Android API for using RCS is open yet. So I don't know if it's possible to implement RCS in quik. It's a difficult question to work on.
> 
> Here is my question: how much money do we want to pay to hire an engineer (software + telecom) to work on it?

I don't think there is anything that would suggest that Google would ever implement an open RCS api to the Android OS. The only realistic option is to implement RCS in the app. That isn't easy either, because the spec is really massive. And, if you want or need USIM authentication, [you will need system privilege](https://github.com/Hirohumi/RustyRcs/issues/1). SMS based authentication should be possible for a regular app, though.

There are at least two open source prototype RCS implementations already existing, the Hirohumi's one, and an older one named [rcsjta](https://github.com/android-rcs/rcsjta). Unfortunately, the former lacks documentation, and the latter has been unmaintained for years, so utilizing either of them wouldn't be a trivial task.

**Comment by The-Nickname** [comment]
~~FYI: RCS support is currently being actively worked on at microG, see https://github.com/microg/GmsCore/pull/2995~~

Not anymore.

**Comment by YannBirba** [comment]
Nice!

> FYI: RCS support is currently being actively worked on at microG, see https://github.com/microg/GmsCore/pull/2995

**Comment by RokeJulianLockhart** [comment]
@The-Nickname, it is not: see https://github.com/microg/GmsCore/pull/2995#pullrequestreview-3102853272. The PR was AI spam.

**Comment by The-Nickname** [comment]
> [@The-Nickname](https://github.com/The-Nickname), it is not: see [microg/GmsCore#2995 (review)](https://github.com/microg/GmsCore/pull/2995#pullrequestreview-3102853272). The PR was AI spam.

Unfortunately, well I didnt look much into it, just thought some ppl might be interested to help.
Well the Bug Bounty is still open: https://github.com/microg/GmsCore/issues/2994, and Marvin doesn't seem to be against it if properly done: https://github.com/microg/GmsCore/pull/2995#issuecomment-3172426911
